volume_type_validation/volume_type_validation.py

```diff
@@ -0,0 +1,187 @@
+"""AWS Volume Type Validation interface."""
+import functools
+import operator
+
+import pandas as pd
+from pluck import pluck
+
+
+class VolumeTypeValidationResult:
```

Comment:
@tumido something like Response object should be there, `dict` is a very poor interface. But yes, the response object is very simple now, so lets rather focus on designing a proper implementation. So response object with `add, validate` method and response item object (something like json schema), so we can use also for generating response documentation and version it.

The benchmark is out of context, since we are talking like 1ms vs 4ms? So in the whole model performance, this will be like 0.01%? For benchmarks, we should start to benchmark the whole model, so e.g. generating mock dataframes like I do at the bottom of https://gitlab.cee.redhat.com/insights-aiops/modeling/blob/master/use_cases/ocp4-aws-instance-type-validation/aws-instance-type-validation.ipynb `In [115]:`

Also lets start to do these reviews for the jupyter notebook, not when we bring it here?

Reply:
I would like to have the code coming to AI-Ops repos reviewed no matter what happened in GitLab. So yeah, we can start reviewing the code there as well, sure.

Well as you said - let's start defining an interface. I agree it would be nice to have some standards. The proposed implementation is not good. Maybe we should start working on that as a part of the new design together?

Anyways, the above is the outcome of my review. I've also offered two options how to change it and clear the clutter. Also having the code mentioning somewhere that it's an interface, would be nice. That way, maybe someone would recognize it. :D This way it's a singleton `dict` wrapper which tries to acts as a `defaultdict`, just more inefficient, sorry.

Reply:
@tumido `review -> copy paste -> review` seems like ineffective process, but sure the code should be looked through just for sure. But the main model review should be done in the jupyter notebook.

Since the `dict` and `defaultdict` should not be way forward, so I don't see any point going back that. `VolumeTypeValidationResult` is a way forward, lets do version 2 of that and continue. So, yes lets design it an do it. Lets try a simple version and continue until it's awesome.

Reply:
I think we should refactor the Jupyter notebook implementation first, to address the above.